src/Task/Task.ts

```diff
@@ -361,31 +361,34 @@ export class Task extends ListItem {
             today,
         );
 
-        let nextOccurrence: Occurrence | null = null;
-        if (newStatus.isCompleted()) {
-            if (!this.status.isCompleted() && this.recurrence !== null) {
-                nextOccurrence = this.recurrence.next(today);
-            }
-        }
-
         const toggledTask = new Task({
             ...this,
             status: newStatus,
             doneDate: newDoneDate,
             cancelledDate: newCancelledDate,
         });
 
-        const newTasks: Task[] = [];
+        if (!newStatus.isCompleted()) {
+            return [toggledTask];
+        }
 
-        if (nextOccurrence !== null) {
-            const nextTask = this.createNextOccurrence(newStatus, nextOccurrence);
-            newTasks.push(nextTask);
+        if (this.status.isCompleted()) {
+            return [toggledTask];
         }
 
-        // Write next occurrence before previous occurrence.
-        newTasks.push(toggledTask);
+        const recurrence = this.recurrence;
+        if (recurrence === null) {
+            return [toggledTask];
+        }
 
-        return newTasks;
+        const nextOccurrence = recurrence.next(today);
+        if (nextOccurrence === null) {
+            return [toggledTask];
+        }
```

Comment:
Thanks. This is definitely an improvement and I will merge it.

I find that the multiple `if` conditions all with the same content requires some thought to figure out the conditions... So I did some experimenting and came up with this as the sort of thing that might make it a bit clearer still... 

```ts
        const originalStatusWasDone = this.status.isCompleted();
        const newStatusIsNotDone = !newStatus.isCompleted();
        const noRecurrenceRule = this.recurrence === null;

        const noNewRecurrence = newStatusIsNotDone || originalStatusWasDone || noRecurrenceRule;
        if (noNewRecurrence) {
            return [toggledTask];
        }

        const nextOccurrence = this.recurrence.next(today);
        if (nextOccurrence === null) {
            return [toggledTask];
        }
```

Reply:
(I'm not wedded to the variable names - I gave them mere seconds of thought, just to do a quick experiment...)

Reply:
I had this thought too but haven’t found a solution as elegant as yours =) you should have pushed this to my branch before merging =)

Reply:
OK I'll do this now. (I hadn't done it safely yesterday, so would not have pushed)

Reply:
> OK I'll do this now. (I hadn't done it safely yesterday, so would not have pushed)

This was the end result:

https://github.com/obsidian-tasks-group/obsidian-tasks/compare/3207040f32d558916dffe8f267e789932745b556...04ecdb518695a616930afb29643b33618492e465

Reply:
Very nice! This looks so much better and easily discoverable (just put add a `!` to any condition in the `if` and will see tests fail =)

I wonder if we can add something like

```ts
const noNextOccurrence = this.recurrence?.next(today);`
```

 and "compress" the 2 conditions into one. But I'm nitpicking here =)

Reply:
I'm not seeing the benefit of that... Feel free to show me next time we meet!